// ng-packagr/lib/steps/tsc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var path = require("path");
var ts = require("typescript");
var typescript_1 = require("typescript");
var fs_1 = require("../util/fs");
var log_1 = require("../util/log");
/**
 * Downlevels a .js file from ES2015 to ES5. Internally, uses `tsc`.
 *
 * @param inputFile Tran
 * @param outputFile
 */
exports.downlevelWithTsc = function (inputFile, outputFile) {
    return Promise.resolve(log_1.debug("tsc " + inputFile + " to " + outputFile))
        .then(function () { return fs_1.readFile(inputFile); })
        .then(function (input) { return ts.transpileModule(trimSourceMap(input.toString()), {
        fileName: path.basename(outputFile),
        moduleName: path.basename(outputFile, '.js'),
        compilerOptions: {
            target: typescript_1.ScriptTarget.ES5,
            module: typescript_1.ModuleKind.ES2015,
            allowJs: true,
            sourceMap: true
        }
    }); })
        .then(function (transpiled) {
        var sourceMap = JSON.parse(transpiled.sourceMapText);
        sourceMap['file'] = path.basename(outputFile);
        sourceMap['sources'] = [path.basename(inputFile)];
        return Promise.all([
            fs_1.writeFile(outputFile, transpiled.outputText),
            fs_1.writeFile(outputFile + ".map", JSON.stringify(sourceMap))
        ]);
    });
};
var REGEXP = /\/\/# sourceMappingURL=.*\.js\.map/;
var trimSourceMap = function (fileContent) {
    if (fileContent.match(REGEXP)) {
        return fileContent.replace(/\/\/# sourceMappingURL=.*\.js\.map/, '');
    }
    else {
        return fileContent;
    }
};
