// ng-packagr/lib/util/fs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs");
var path = require("path");
var read = require('read-file');
var promisify_1 = require("./promisify");
exports.readFile = function (file) {
    return promisify_1.promisify(function (resolveOrReject) {
        var fileReadOptions = {
            encoding: 'utf8',
            normalize: true
        };
        read(file, fileReadOptions, function (err, buffer) {
            if (buffer) {
                resolveOrReject(err, buffer.toString());
            }
            else {
                resolveOrReject(err);
            }
        });
    });
};
exports.writeFile = function (file, content) {
    return promisify_1.promisify(function (resolveOrReject) {
        fs.writeFile(file, content, resolveOrReject);
    });
};
exports.findFromDirectory = function (dir, file, cb) {
    var fileName = path.resolve(dir, file);
    fs.access(fileName, fs.constants.R_OK, function (err) {
        if (err) {
            exports.findFromDirectory(path.resolve(dir, '..'), file, cb);
        }
        else {
            cb(fileName);
        }
    });
};
